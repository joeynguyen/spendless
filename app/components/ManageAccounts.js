import React, { Component, PropTypes } from 'react';
import { Modal, Button, ButtonInput, Collapse, Panel, Input } from 'react-bootstrap';
import ApiKey from '../api-key.js';

// PouchDB is loaded externally through a script tag in the browser
const db = new PouchDB('accounts');
let remoteCouch = false;
if (ApiKey.user !== '' && ApiKey.pass !== '') {
  remoteCouch = `http://${ApiKey.user}:${ApiKey.pass}@joeynguyen.cloudant.com/spendless-accounts`;
}

export default class ManageAccountsWindow extends Component {
  static propTypes = {
    onNewAccount: PropTypes.func.isRequired,
    showModal: PropTypes.bool.isRequired,
    close: PropTypes.func.isRequired,
  }
  state = {
    accountName: '',
    accountType: '',
    accountCompany: '',
    showAdd: false,
    showBank: false,
    showCredit: false
  }
  syncDB() {
    const opts = {live: true};
    db.sync(remoteCouch, opts)
      .on('complete', function(success) {
        console.log('CouchDB sync success :', success);
      })
      .on('error', function(err) {
        console.log('CouchDB sync error :', err);
      });
  }
  toggleAddAccount = () => {
    this.setState({showAdd: !this.state.showAdd});
  }
  closeWindow = () => {
    this.setState({showAdd: false, showBank: false, showCredit: false});
    this.props.close();
  }
  handleAccountNameChange = (e) => {
    this.setState({accountName: e.target.value});
  }
  handleAccountTypeChange = (e) => {
    if (e.target.value === 'creditcard') {
      this.setState({showCredit: true, showBank: false, accountCompany: ''});
    } else if (e.target.value === 'bank') {
      this.setState({showBank: true, showCredit: false, accountCompany: ''});
    } else {
      this.setState({showCredit: false, showBank: false});
    }
    this.setState({accountType: e.target.value});
  }
  handleAccountCompanyChange = (e) => {
    this.setState({accountCompany: e.target.value});
  }
  handleSubmit = (e) => {
    e.preventDefault();
    const newAccount = {
      '_id': new Date().toISOString(),
      'name': this.state.accountName,
      'type': this.state.accountType,
      'company': this.state.accountCompany,
    };
    const self = this;
    console.log(newAccount);
    db.put(newAccount).then(function(result) {
      console.log('Successfully added new account');
      console.log(result);
      self.updateAccounts(newAccount);
      self.setState({showAdd: false, showBank: false, showCredit: false});
      // TODO: reset form inputs
      // TODO: Add success message after successful submit
    }).catch(function(err) {
      console.log(err);
      // TODO: Add error message after submit fail
    });
    if (remoteCouch) {
      this.syncDB();
    }
  }
  updateAccounts = (newAccount) => {
    this.props.onNewAccount(newAccount);
  }
  render() {
    let addButton = {};
    if (this.state.showAdd) {
      addButton = { style: 'danger', class: '', text: 'Cancel' };
    } else {
      addButton = { style: 'primary', class: 'fa fa-plus', text: ' Add Account' };
    }
    return (
      <Modal show={this.props.showModal} onHide={this.closeWindow}>
        <Modal.Header closeButton>
          <Modal.Title>Manage Accounts</Modal.Title>
        </Modal.Header>
        <Modal.Body>
          <div className="row">
            <div className="col-xs-8">
              <Button onClick={this.toggleAddAccount} bsStyle={addButton.style} bsSize="large" block><i className={addButton.class}></i>{addButton.text}</Button>
              <Panel collapsible expanded={this.state.showAdd}>
                {/* TODO: Add form validation. Don't allow 'select' value to be chosen */}
                <form onSubmit={this.handleSubmit}>
                  <Input type="text" label="Name" placeholder="Enter a name for the account" value={this.state.accountName} onChange={this.handleAccountNameChange} />
                  <Input type="select" label="Type" placeholder="Type" value={this.state.accountType} onChange={this.handleAccountTypeChange} >
                    <option value="select">select</option>
                    <option value="bank">Bank</option>
                    <option value="creditcard">Credit Card</option>
                  </Input>

                  <Collapse in={this.state.showBank}>
                    <div>
                      <Input type="text" label="Name of Institution" placeholder="Enter the name of the financial institution" value={this.state.accountCompany} onChange={this.handleAccountCompanyChange } />
                    </div>
                  </Collapse>
                  <Collapse in={this.state.showCredit}>
                    <div>
                      <Input type="select" label="Credit Card Company" placeholder="Credit Card Company" value={this.state.accountCompany} onChange={this.handleAccountCompanyChange}>
                        <option value="select">select</option>
                        <option value="Visa">Visa</option>
                        <option value="MasterCard">Mastercard</option>
                        <option value="American Express">American Express</option>
                        <option value="Discover">Discover</option>
                        <option value="Diners Club">Diners Club</option>
                        <option value="JCB">JCB</option>
                        <option value="Other">Other</option>
                      </Input>
                    </div>
                  </Collapse>

                  <ButtonInput bsStyle="primary" type="submit" value="Add" />
                </form>
              </Panel>
              <hr />
              {/* Use state to display 'No accounts' */}
              <p>No accounts found.</p>
            </div>

            <div className="col-xs-4">
              <ul>
                <li>1 Checking accounts</li>
                <li>2 Credit card accounts</li>
              </ul>
            </div>
          </div>

        </Modal.Body>
        <Modal.Footer>
          <Button onClick={this.closeWindow}>Close</Button>
        </Modal.Footer>
      </Modal>
    );
  }
}
