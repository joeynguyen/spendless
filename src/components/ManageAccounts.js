import React, { Component, PropTypes } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { Modal, Button } from 'react-bootstrap';
import { toggleManageAccounts } from '../manage-accounts/ManageAccountsActions.js';
import AddAccountContainer from '../manage-accounts/AddAccountContainer.js';

class ManageAccountsWindow extends Component {
  static propTypes = {
    showManageAccounts: PropTypes.bool.isRequired,
    doToggleManageAccounts: PropTypes.func.isRequired
  }
  closeWindow = () => {
    // this.setState({showAdd: false, showBank: false, showCredit: false});
    this.props.doToggleManageAccounts(false);
  }
  render() {
    return (
      <Modal show={this.props.showManageAccounts} onHide={this.closeWindow}>
        <Modal.Header closeButton>
          <Modal.Title>Manage Accounts</Modal.Title>
        </Modal.Header>
        <Modal.Body>
          <div className="row">
            <div className="col-xs-8">
              <AddAccountContainer />
              <hr />
              {/* Use state to display 'No accounts' */}
              {/* TODO: Show all accounts in this window */}
              <p>No accounts found.</p>
            </div>

            <div className="col-xs-4">
              <ul>
                {/* TODO: Dynamically show number of accounts */}
                <li>1 Checking accounts</li>
                <li>2 Credit card accounts</li>
              </ul>
            </div>
          </div>

        </Modal.Body>
        <Modal.Footer>
          <Button onClick={this.closeWindow}>Close</Button>
        </Modal.Footer>
      </Modal>
    );
  }
}

function mapStateToProps(state) {
  return {
    showManageAccounts: state.showManageAccounts
  };
}

function mapDispatchToProps(dispatch) {
  return bindActionCreators({ doToggleManageAccounts: toggleManageAccounts }, dispatch);
}

export default connect(mapStateToProps, mapDispatchToProps)(ManageAccountsWindow);
