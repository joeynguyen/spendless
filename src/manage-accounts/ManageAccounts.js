import React, { PropTypes } from 'react';
import { Modal, Button } from 'react-bootstrap';
import AddAccountContainer from './AddAccountContainer.js';
import AccountsList from './AccountsList.js';

const ManageAccounts = ({ manageAccountsVisible, doToggleManageAccounts, accounts }) => {
  const ccAccounts = accounts.filter(account => account.type === 'creditcard');
  const bankAccounts = accounts.filter(account => account.type === 'bank');
  return (
    <Modal
      show={manageAccountsVisible}
      onHide={doToggleManageAccounts}
    >
      <Modal.Header closeButton>
        <Modal.Title>Manage Accounts</Modal.Title>
      </Modal.Header>
      <Modal.Body>
        <div className="row">
          <div className="col-xs-8">
            <AddAccountContainer />
            <hr />
            <AccountsList accounts={accounts} />
          </div>

          <div className="col-xs-4">
            <ul>
              <li>{bankAccounts.length} Bank accounts</li>
              <li>{ccAccounts.length} Credit card accounts</li>
            </ul>
          </div>
        </div>

      </Modal.Body>
      <Modal.Footer>
        <Button onClick={doToggleManageAccounts}>Close</Button>
      </Modal.Footer>
    </Modal>
  );
};
ManageAccounts.propTypes = {
  manageAccountsVisible: PropTypes.bool.isRequired,
  doToggleManageAccounts: PropTypes.func.isRequired,
  accounts: PropTypes.array.isRequired,
};

export default ManageAccounts;
