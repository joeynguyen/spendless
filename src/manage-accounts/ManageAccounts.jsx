import PropTypes from 'prop-types';
import React from 'react';
import { Col, Row, Button, Modal } from 'antd';
import AddAccount from './AddAccount.jsx';
import AccountsList from './AccountsList.jsx';

const ManageAccounts = ({ accounts, actions, manageAccountsVisible }) => {
  const ccAccounts = accounts.filter(account => account.type === 'creditcard');
  const bankAccounts = accounts.filter(account => account.type === 'bank');
  const modalFooter = [(
    <Button
      key={0}
      size="large"
      onClick={actions.toggleManageAccounts}
    >Close</Button>
  )];
  return (
    <Modal
      closable
      maskClosable={false}
      visible={manageAccountsVisible}
      onCancel={actions.toggleManageAccounts}
      title="Manage Accounts"
      footer={modalFooter}
    >
      <Row>
        <Col span={16}>
          <AddAccount />
          <AccountsList accounts={accounts} />
        </Col>

        <Col span={8} style={{padding: '0 12px' }}>
          <ul>
            <li>{bankAccounts.length} Bank accounts</li>
            <li>{ccAccounts.length} Credit card accounts</li>
          </ul>
        </Col>
      </Row>
    </Modal>
  );
};
ManageAccounts.propTypes = {
  accounts: PropTypes.array.isRequired,
  actions: PropTypes.object.isRequired,
  manageAccountsVisible: PropTypes.bool.isRequired,
};

export default ManageAccounts;
