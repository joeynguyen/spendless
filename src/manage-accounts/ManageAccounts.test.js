import expect from 'expect';
import React from 'react';
import { shallow } from 'enzyme';
import ManageAccounts from './ManageAccounts.jsx';

const myProps = {
  accounts: [
    {
      _id: '2016-04-24T20:47:37.374Z',
      _rev: '2-8c6e36eddfb9990c199546ad267cfd61',
      company: 'Santander Banks',
      name: 'Santander Checkings',
      type: 'bank',
    },
    {
      _id: '2016-02-24T07:01:24.948Z',
      _rev: '13-2bfa9371e269957dcbf2946c1cdd6cd5',
      company: 'MasterCard',
      name: 'Chase Freedom',
      type: 'creditcard',
    },
    {
      _id: '2016-03-24T08:10:58.232Z',
      _rev: '14-57184e4f5b50323eeeb6a919faf1af48',
      company: 'Bank of America',
      name: 'BofA Checkings',
      type: 'bank',
    },
  ],
  actions: {},
  manageAccountsVisible: true,
};

function setup(props) {
  return shallow(<ManageAccounts {...props} />);
}

describe('ManageAccounts', () => {
  const component = setup(myProps);

  it('has an AddAccount component', done => {
    expect(component.find('AddAccount').length).toBe(1);
    done();
  });

  it('has a AccountsList component', done => {
    expect(component.find('AccountsList').length).toBe(1);
    done();
  });

  it('AccountsList has an accounts prop with list of accounts passed to it', done => {
    expect(component.find('AccountsList').props().accounts).toExist();
    expect(component.find('AccountsList').props().accounts).toBe(
      myProps.accounts
    );
    done();
  });

  it('has a close modal button', done => {
    expect(component.find('Button[name="close-manage-accounts"]').length).toBe(
      1
    );
    done();
  });
});
