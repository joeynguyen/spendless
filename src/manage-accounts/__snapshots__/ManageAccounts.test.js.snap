// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ManageAccounts has an ManageAccounts component with props 1`] = `
<Modal
  closable={true}
  confirmLoading={false}
  footer={
    Array [
      <Button
        ghost={false}
        loading={false}
        onClick={undefined}
        prefixCls="ant-btn"
        size="large"
      >
        Close
      </Button>,
    ]
  }
  id="ManageAccounts"
  maskClosable={false}
  maskTransitionName="fade"
  okType="primary"
  prefixCls="ant-modal"
  title="Manage Accounts"
  transitionName="zoom"
  visible={true}
  width={520}
>
  <Row
    gutter={0}
  >
    <Col
      span={16}
    >
      <AddAccount />
      <AccountsList
        accounts={
          Array [
            Object {
              "_id": "2016-04-24T20:47:37.374Z",
              "_rev": "2-8c6e36eddfb9990c199546ad267cfd61",
              "company": "Santander Banks",
              "name": "Santander Checkings",
              "type": "bank",
            },
            Object {
              "_id": "2016-02-24T07:01:24.948Z",
              "_rev": "13-2bfa9371e269957dcbf2946c1cdd6cd5",
              "company": "MasterCard",
              "name": "Chase Freedom",
              "type": "creditcard",
            },
            Object {
              "_id": "2016-03-24T08:10:58.232Z",
              "_rev": "14-57184e4f5b50323eeeb6a919faf1af48",
              "company": "Bank of America",
              "name": "BofA Checkings",
              "type": "bank",
            },
          ]
        }
      />
    </Col>
    <Col
      span={8}
      style={
        Object {
          "padding": "0 12px",
        }
      }
    >
      <ul>
        <li>
          2
           Bank accounts
        </li>
        <li>
          1
           Credit card accounts
        </li>
      </ul>
    </Col>
  </Row>
</Modal>
`;
